src/field.rs

```diff
@@ -111,32 +151,92 @@ pub trait FieldElement:
+    /// sequence of `Self::ENCODING.size`-byte sequences. Each sub-sequence is interpreted in
+    /// `Self::ENCODING.order` order.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the length of the provided byte slice is not a multiple of the size of a
+    /// field element, or if any of the values in the byte slice are invalid encodings of a field
+    /// element as documented in `FieldElement::try_from_reader`.
+    ///
+    /// # Notes
+    ///
+    /// Ideally we would implement `From<&[u8]> for Vec<F: FieldElement>` or the corresponding
+    /// `Into`, but the orphan rule and the stdlib's blanket implementations of `Into` make this
+    /// impossible.
+    fn byte_slice_into_vec(bytes: &[u8]) -> Result<Vec<Self>, FieldError> {
+        if bytes.len() % Self::ENCODING.size != 0 {
+            return Err(FieldError::ShortRead);
+        }
+        let mut vec = Vec::with_capacity(bytes.len() / Self::ENCODING.size);
+        for chunk in bytes.chunks_exact(Self::ENCODING.size) {
+            vec.push(Self::try_from_reader(&mut Cursor::new(chunk))?);
+        }
+        Ok(vec)
+    }
 }
 
 macro_rules! make_field {
     (
         $(#[$meta:meta])*
-        $elem:ident, $int:ident, $fp:ident, $bytes:literal
+        $elem:ident, $int:ident, $fp:ident, $bytes:literal, $endian:expr,
     ) => {
         $(#[$meta])*
         #[derive(Clone, Copy, PartialOrd, Ord, Default, Deserialize, Serialize)]
         pub struct $elem(u128);
 
         impl $elem {
+            /// Attempts to instantiate an `$elem` from the first `Self::ENCODING.size` bytes in
+            /// the provided slice, interpreted in `Self::ENCODING.order` order. The decoded value
+            /// will be bitwise-ANDed with `mask` before reducing it using the field modulus.
+            ///
+            /// # Errors
+            ///
+            /// An error is returned if the provided slice is not long enough to encode a field
+            /// element or if the decoded value is greater than the field prime.
+            ///
+            /// # Notes
+            ///
+            /// We cannot use `u128::from_le_bytes` here because that function expects inputs
+            /// to be exactly 16 bytes long. Our encoding of most field elements is more compact,
+            /// and does not always correspond to the size of an integer type (specifically,
+            /// `Field80`'s encoding is 10 bytes, even though it is a 16 byte `u128` in memory).
             fn try_from_bytes(bytes: &[u8], mask: u128) -> Result<Self, FieldError> {
-                if Self::BYTES > bytes.len() {
-                    return Err(FieldError::FromBytesShortRead);
+                if Self::ENCODING.size > bytes.len() {
+                    return Err(FieldError::ShortRead);
                 }
 
                 let mut int = 0;
-                for i in 0..Self::BYTES {
+                for mut i in 0..Self::ENCODING.size {
```

Comment:
This code has a bug. Not only are you reversing the index into `bytes`, you're also changing the number of bits you are rotating the input by `(i << 3)`. The result is that `Self::ENCODING.order` is not being respected and you're always doing little endian. To fix this, change this loop to
```
                 for i in 0..Self::ENCODING.size {
                    let j = match Self::ENCODING.order {
                        ByteOrder::LittleEndian => i,
                        ByteOrder::BigEndian => Self::ENCODING.size - i - 1,
                    };
               
                    int |= (bytes[j] as u128) << (i << 3);
                }
```
Similarly for `impl From<$elem> for [u8; $elem::ENCODING.size]` below.

While at it, let's consider adding unit tests for checking the byte order is correct.